File: tday-backend/src/main/kotlin/com/ohmz/tday/services/MemoryCache.kt
package com.ohmz.tday.services

import java.util.concurrent.ConcurrentHashMap

data class CacheEntry<T>(val data: T, val expiresAt: Long)

interface CacheService {
    fun <T> get(key: String): T?
    fun <T> set(key: String, data: T, ttlMs: Long = 60_000L)
    fun invalidateForUser(userId: String)
    fun invalidateUserEndpoint(userId: String, endpoint: String)
    fun clear()
    fun cacheKey(userId: String, endpoint: String, params: Map<String, String>? = null): String
    fun invalidateTodoCaches(userId: String)
    fun invalidateListCaches(userId: String)
    fun invalidateCompletedCaches(userId: String)
}

class CacheServiceImpl : CacheService {
    private val defaultTtlMs = 60_000L
    private val cleanupIntervalMs = 5 * 60_000L
    private val store = ConcurrentHashMap<String, CacheEntry<Any?>>()
    private var lastCleanup = System.currentTimeMillis()

    @Suppress("UNCHECKED_CAST")
    override fun <T> get(key: String): T? {
        lazyCleanup()
        val entry = store[key] ?: return null
        if (System.currentTimeMillis() > entry.expiresAt) {
            store.remove(key)
            return null
        }
        return entry.data as? T
    }

    override fun <T> set(key: String, data: T, ttlMs: Long) {
        store[key] = CacheEntry(data, System.currentTimeMillis() + ttlMs)
    }

    override fun invalidateForUser(userId: String) {
        val prefix = "$userId:"
        store.keys.filter { it.startsWith(prefix) }.forEach { store.remove(it) }
    }

    override fun invalidateUserEndpoint(userId: String, endpoint: String) {
        val prefix = "$userId:$endpoint"
        store.keys.filter { it.startsWith(prefix) }.forEach { store.remove(it) }
    }

    override fun clear() = store.clear()

    override fun cacheKey(userId: String, endpoint: String, params: Map<String, String>?): String {
        if (params.isNullOrEmpty()) return "$userId:$endpoint"
        val sorted = params.entries.sortedBy { it.key }.joinToString("&") { "${it.key}=${it.value}" }
        return "$userId:$endpoint:$sorted"
    }

    override fun invalidateTodoCaches(userId: String) {
        invalidateUserEndpoint(userId, "todo")
        invalidateUserEndpoint(userId, "completedTodo")
    }

    override fun invalidateListCaches(userId: String) {
        invalidateUserEndpoint(userId, "list")
        invalidateUserEndpoint(userId, "todo")
    }

    override fun invalidateCompletedCaches(userId: String) {
        invalidateUserEndpoint(userId, "completedTodo")
    }

    private fun lazyCleanup() {
        val now = System.currentTimeMillis()
        if (now - lastCleanup < cleanupIntervalMs) return
        lastCleanup = now
        store.entries.removeIf { now > it.value.expiresAt }
    }
}

File: tday-web/src/lib/cache/memoryCache.ts
interface CacheEntry<T> {
  data: T;
  expiresAt: number;
}

const DEFAULT_TTL_MS = 60_000;
const CLEANUP_INTERVAL_MS = 5 * 60_000;

class MemoryCache {
  private store = new Map<string, CacheEntry<unknown>>();
  private lastCleanup = Date.now();

  get<T>(key: string): T | undefined {
    this.lazyCleanup();
    const entry = this.store.get(key);
    if (!entry) return undefined;
    if (Date.now() > entry.expiresAt) {
      this.store.delete(key);
      return undefined;
    }
    return entry.data as T;
  }

  set<T>(key: string, data: T, ttlMs = DEFAULT_TTL_MS): void {
    this.store.set(key, {
      data,
      expiresAt: Date.now() + ttlMs,
    });
  }

  invalidateForUser(userId: string): void {
    const prefix = `${userId}:`;
    for (const key of this.store.keys()) {
      if (key.startsWith(prefix)) {
        this.store.delete(key);
      }
    }
  }

  invalidateUserEndpoint(userId: string, endpoint: string): void {
    const prefix = `${userId}:${endpoint}`;
    for (const key of this.store.keys()) {
      if (key.startsWith(prefix)) {
        this.store.delete(key);
      }
    }
  }

  clear(): void {
    this.store.clear();
  }

  get size(): number {
    return this.store.size;
  }

  private lazyCleanup() {
    const now = Date.now();
    if (now - this.lastCleanup < CLEANUP_INTERVAL_MS) return;
    this.lastCleanup = now;
    for (const [key, entry] of this.store) {
      if (now > entry.expiresAt) {
        this.store.delete(key);
      }
    }
  }
}

const globalCache = globalThis as unknown as { __apiCache?: MemoryCache };
export const apiCache: MemoryCache =
  globalCache.__apiCache ?? (globalCache.__apiCache = new MemoryCache());

export function cacheKey(
  userId: string,
  endpoint: string,
  params?: Record<string, string>,
): string {
  if (!params || Object.keys(params).length === 0) {
    return `${userId}:${endpoint}`;
  }
  const sorted = Object.entries(params)
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([k, v]) => `${k}=${v}`)
    .join("&");
  return `${userId}:${endpoint}:${sorted}`;
}

export function invalidateTodoCaches(userId: string): void {
  apiCache.invalidateUserEndpoint(userId, "todo");
  apiCache.invalidateUserEndpoint(userId, "completedTodo");
}

export function invalidateListCaches(userId: string): void {
  apiCache.invalidateUserEndpoint(userId, "list");
  apiCache.invalidateUserEndpoint(userId, "todo");
}

export function invalidateCompletedCaches(userId: string): void {
  apiCache.invalidateUserEndpoint(userId, "completedTodo");
}

export function invalidateSettingsCaches(userId: string): void {
  apiCache.invalidateUserEndpoint(userId, "app-settings");
}
